libc/utils/HdrGen/PublicAPICommand.cpp

```diff
@@ -87,6 +89,94 @@ void writeAPIFromIndex(APIIndexer &G,
+      MacroAttr[Attr->getValueAsString("Macro")] = Attr;
+    }
+  }
+
+  auto GetStyle = [](llvm::Record *Instance) {
+    auto Style = Instance->getValueAsString("Style");
+    if (Style == "cxx11")
+      return AttributeStyle::Cxx11;
+    if (Style == "gnu")
+      return AttributeStyle::Gnu;
+    return AttributeStyle::Declspec;
+  };
+
+  auto GetNamespace = [](llvm::Record *Instance) {
+    auto Namespace = Instance->getValueAsString("Namespace");
+    // Empty namespace is likely to be most standard-compliant.
+    if (Namespace.empty())
+      return AttributeNamespace::None;
+    // Dispatch clang version before gnu version.
+    if (Namespace == "clang")
+      return AttributeNamespace::Clang;
+    return AttributeNamespace::Gnu;
+  };
+
+  for (auto &[Macro, Attr] : MacroAttr) {
+    auto Instances = Attr->getValueAsListOfDefs("Instances");
+    llvm::SmallVector<std::pair<AttributeStyle, llvm::Record *>> Styles;
+    std::transform(Instances.begin(), Instances.end(),
+                   std::back_inserter(Styles),
+                   [&](llvm::Record *Instance)
+                       -> std::pair<AttributeStyle, llvm::Record *> {
+                     auto Style = GetStyle(Instance);
+                     return {Style, Instance};
+                   });
+    // Effectively sort on the first field
+    std::sort(Styles.begin(), Styles.end(), [&](auto &a, auto &b) {
+      if (a.first == AttributeStyle::Cxx11 && b.first == AttributeStyle::Cxx11)
+        return GetNamespace(a.second) < GetNamespace(b.second);
+      return a.first < b.first;
+    });
+    for (auto &[Style, Instance] : Styles) {
+      if (Style == AttributeStyle::Cxx11) {
+        OS << "#if !defined(" << Macro << ") && defined(__cplusplus)";
+        auto Namespace = GetNamespace(Instance);
+        if (Namespace == AttributeNamespace::Clang)
+          OS << " && defined(__clang__)\n";
+        else if (Namespace == AttributeNamespace::Gnu)
+          OS << " && defined(__GNUC__)\n";
+        else
+          OS << '\n';
+        OS << "#define " << Macro << " [[";
+        if (Namespace == AttributeNamespace::Clang)
+          OS << "clang::";
+        else if (Namespace == AttributeNamespace::Gnu)
+          OS << "gnu::";
+        OS << Instance->getValueAsString("Attr") << "]]\n";
+        OS << "#endif\n";
+      }
+      if (Style == AttributeStyle::Gnu) {
+        OS << "#if !defined(" << Macro << ") && defined(__GNUC__)\n";
+        OS << "#define " << Macro << " __attribute__((";
+        OS << Instance->getValueAsString("Attr") << "))\n";
+        OS << "#endif\n";
+      }
+      if (Style == AttributeStyle::Declspec) {
+        OS << "#if !defined(" << Macro << ") && defined(_MSC_VER)\n";
+        OS << "#define " << Macro << " __declspec(";
+        OS << Instance->getValueAsString("Attr") << ")\n";
+        OS << "#endif\n";
+      }
+    }
+    OS << "#if !defined(" << Macro << ")\n";
+    OS << "#define " << Macro << '\n';
+    OS << "#endif\n";
```

Comment:
I'd be curious to see examples of the resulting output.

I love function attributes, but using them portably is somewhat a PITA.

`__has_attribute` helps a lot, if your compiler supports it.  If not, you have to rely on compiler version checks (or omit the attributes or explicitly not support those compiler versions).

So if we expose these on the client facing headers, we need to ensure that the result works with a wide range of compiler versions.  Perhaps wider than we require to build llvm libc.  I don't think we make the distinction in our docs, but we probably should.

Reply:
We already require pretty modern GCC / Clang to compile. Worst case we can do `defined(__has_attribute) && __has_attribute` patterns where the default is nothing.

Reply:
Gcc saids:
```
 In C code, if compiling for strict conformance to standards before C23, operand must be a valid identifier. Otherwise, operand may be optionally introduced by the attribute-scope:: prefix.
 ```
Clang seems not be able to parse `__has_attribute(gnu::const)`.
As a result, we may not want to use `__has_attribute` as it is not applicable in many cases.

Reply:
> We already require pretty modern GCC / Clang to compile.

Right, but if we expose this headers contents to userspace, then that might be a different set of compilers to support.

> Clang seems not be able to parse __has_attribute(gnu::const).

You don't use the `gnu::` in the test.
https://godbolt.org/z/dP8Yj6PTM

Reply:
> I'd be curious to see examples of the resulting output.

Bump. Can you share how this changes the generated header declaration for `fabs`?  That might be nice to include in the commit message even.